File: game-frontend/src/Model/tile.ts

export enum Block {
  HOUSE, FONTAIN, CIRCUS, WIND_TURBINE, NOTHING
}

export enum Type {
  TREE, WATER, GRASS
}

export class Tile {

  private occupied: boolean;
  type: Type;
  block: Block ;
  private x: number;
  private y: number;
  constructor(  x: number, y: number) {
    this.x = x;
    this.y = y;
    this.type = this.getRandomType();
    this.occupied = (this.type !== Type.GRASS);
    this.block = Block.NOTHING;
  }
  getRandomType(): Type{
    const random = Math.floor(Math.random() * Math.floor(3));
    switch (random) {
      case 0 :
        return Type.GRASS;
      case 1 :
        return Type.TREE;
      default:
        return Type.WATER;
    }
  }
  equals(obj: Tile): boolean {
    return this.x === obj.x && this.y === obj.y && this.type === obj.type;
  }

  get X(): number {
    return this.x;
  }
  set X(value: number) {
    this.x = value;
  }

  get Y(): number{
    return this.y;
  }
  set Y(value: number) {
    this.y = value;
  }

  get Type(): Type{
    return this.type;
  }
  set Type(value: Type) {
    this.x = value;
  }

  get Occupied(): boolean{
    return this.occupied;
  }
  set Occupied(value: boolean) {
    this.occupied = value;
  }

  get Block(): Block{
    return this.block;
  }
  set Block(value: Block) {
    this.block = value;
  }
  radiusEffect(b: Block): number {
    switch (b){
      case Block.CIRCUS: return 3;
      case Block.NOTHING: return 0;
      default: return 1;
    }
  }
  jsonToBlock(b: any): Block {
    switch (b) {
      case 'HOUSE': return Block.HOUSE;
      case 'WIND_TURBINE': return Block.WIND_TURBINE;
      case 'FONTAIN': return Block.FONTAIN;
      case 'CIRCUS' : return Block.CIRCUS;
      default : return Block.NOTHING;
    }
  }
  jsonToType(t: any): Type {
    switch (t) {
      case 'TREE': return Type.TREE;
      case 'WATER': return Type.WATER;
      default: return Type.GRASS;
    }
  }
  jsonToTile(tile: any): Tile {
    const res = new Tile(tile.x, tile.y);
    res.occupied = tile.occupied;
    res.type = this.jsonToType(tile.type);
    res.block = this.jsonToBlock(tile.block);
    return res;
  }

  jsonToTiles(tiles: any): Array<Tile> {
    const liste = [];
    tiles.forEach( t =>
    liste.push(this.jsonToTile(t)));
    return liste;
  }
}

File: game-backend/src/main/java/game/model/Tile.java
package game.model;


import java.util.Random;

public class Tile {
    private  boolean occupied;
    private final int x;
    private final int y;
    private final Type type;
    private Block block;

    /**
     *
     * @param x the coordinate x of the Tile
     * @param y the coordinate y of the Tile
     * @param t Type of the Tile, if it is something else than GRASS, the Tyle is set occupied
     * initially the block NOTHING is on all the Tiles ( help to compute score)
     */
    public Tile(final int x, final int y, final Type t) {
        this.x = x;
        this.y = y;
        this.type = t;
        occupied = (type != Type.GRASS);
        block = Block.NOTHING;

    }
    public Tile() {
        final Random r = new Random();
        x = r.nextInt(10);
        y = r.nextInt(10);
        type = getRandomType();
        occupied = type != Type.GRASS;
        block = Block.NOTHING;
    }

    public void setBlock(final Block b) {
        if(type == Type.GRASS) {
            block = b;
        }
    }

    public Block getBlock() {
        return block;
    }
    public boolean getOccupied() {
        return occupied;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public void setOccupied(final boolean b) {
        occupied = b;
    }

    public Type getType() {
        return type;
    }

    /**
     *
     * @return a random Type for a Tile
     */
    public static Type getRandomType() {
        final Random rand = new Random();
        final int r = rand.nextInt(3);
        switch (r) {
            case 0 :
                return Type.GRASS;
            case 1 :
                return Type.TREE;
            default:
                return Type.WATER;
        }
    }

    @Override
    public boolean equals(final Object obj) {
        if (obj instanceof Tile) {
            return x == ((Tile) obj).x && y == ((Tile) obj).y && type == ((Tile) obj).type;
        } else {
            return false;
        }
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }
}

